### DISCUSSION TITLE
Very slow HTTP img fetching with QLogic 10GbE NIC (cLOM8214, 1077:8020)

### DISCUSSION THREAD

**Discussion by srstsavage** [ipxe/ipxe#548] [Q&A]
Hi all,

I'm using iPXE in a chainloading setup and fetching boot images from an HTTP server (thank you, it's awesome!). I've noticed that the boot images are taking **much** longer to download on servers with QLogic 10GbE NICs (cLOM8214) than other servers with Broadcom NICs.

For a [10MB test file](https://files.axds.co/ipxetest), the Broadcom servers can download the image in less than a second, while the QLogic servers take about 12-13 seconds. Also, the QLogic servers can download the image file at a normal speed once the OS is booted using `curl`, so the problem seems isolated to iPXE.

`ifstat` reports a lot of "socket not connected" errors on the QLogic servers after download:

```
net0:g00:0e:1e:0d:76:b8:using:undionly.on 0000:01:00.0 (Ethernet) [open]
iP[Link:up, TX:208402 TXE:1 RX:1228720 RXE:599]                        
  [TXE: 1 x "Network unreachable (http://ipxe.org/28086011)"]                                                    
  [RXE: 2 x "Invalid argument (http://ipxe.org/1c196002)"]                                      
  [RXE: 574 x "The socket is not connected (http://ipxe.org/380f6001)"]
  [RXE: 20 x "Invalid argument (http://ipxe.org/1c056002)"]                        
  [RXE: 1 x "The socket is not connected (http://ipxe.org/380a6001)"]
```

The QLogic downloads also hang around 98 or 99% for a long time before completing.

![slow_dl](https://user-images.githubusercontent.com/1138892/148339092-3434a51c-b941-4ff3-bc82-1bb150e3a804.gif)

I've attached packet capture examples for both the QLogic and Broadcom servers here:

[packet_captures.zip](https://github.com/ipxe/ipxe/files/7819927/packet_captures.zip)

Happy to provide other debugging information if needed (wasn't sure which driver to use for `DEBUG`). I can see [here](https://ipxe.org/appnote/hardware_drivers) that that Broadcom NICs are supported by the tg3 driver but the only QLogic mention I see is for infiniband hardware...maybe that's the problem?

`lspci -nn` output for the problem QLogic NICs:

```
01:00.0 Ethernet controller [0200]: QLogic Corp. cLOM8214 1/10GbE Controller [1077:8020] (rev 58)
```

and for working Broadcom NICs:

```
04:00.0 Ethernet controller [0200]: Broadcom Inc. and subsidiaries NetXtreme II BCM57810 10 Gigabit Ethernet [14e4:168e] (rev 10)
```

```
02:00.0 Ethernet controller [0200]: Broadcom Inc. and subsidiaries NetXtreme BCM5720 2-port Gigabit Ethernet PCIe [14e4:165f]
```

```
01:00.0 Ethernet controller [0200]: Broadcom Limited NetXtreme II BCM5709S Gigabit Ethernet [14e4:163a] (rev 20)
```

**Comment by NiKiZe** [comment]
Have you verified the link speed? Check the port in the switch.
You are using undionly here, which is th NIC ROM, have you looked for updated ROM/firmware for the NIC?

**Comment by srstsavage** [comment]
Thanks for your reply. Yes, should have mentioned that I have verified the link speed to be 10000Mb/s on both the switch (during iPXE booting and after OS boots) and the NIC via `ethtool` (after OS boots).

In addition to `undionly.kpxe` I have also tried `ipxe.pxe` and `undionly.kkpxe` with the same result.

NIC firmware should be up to date as I've run Dell's `dsu` command to update firmware and watched it apply updates to the NIC.

**Reply by NiKiZe** [reply]
I have not looked at your packet capture, but verify that the server don't do chunked transfer and that content-length header is in the reply.

You might also want to test to force the link to gigabit speed.
Using ipxe.pxe what is the driver used?
(Generally don't use kkpxe builds)

**Comment by srstsavage** [comment]
> Using ipxe.pxe what is the driver used?

undipci

> You might also want to test to force the link to gigabit speed.

Tried this but it didn't seem to affect the results

> verify that the server don't do chunked transfer and that content-length header is in the reply.

Just starting to dig into this. I've been using nginx to serve the image files and it does seem to want to use chunked transfer by default and omit the Content-Length header. Interestingly, when I explicitly disable chunked transfer using `chunked_transfer_encoding off` the transfer still completes by the time _increases_ from about 12s to about 21s. Also the download percentage jumps immediately to 100% and hangs there until the download is complete.

I'll look more into the headers and try some other HTTP servers. Not sure how this would affect the QLogic NICs specifically but I'll keep poking.

**Comment by srstsavage** [comment]
Trying the same file with Python's SimpleHTTP server (no chunked encoding, Content-Length set) results in the same transfer time, although the download percentage progress is now accurate.

```
< HTTP/1.0 200 OK
< Server: SimpleHTTP/0.6 Python/3.10.1
< Date: Fri, 07 Jan 2022 07:26:36 GMT
< Content-type: application/octet-stream
< Content-Length: 10485760
< Last-Modified: Fri, 31 Dec 2021 16:00:58 GMT
```

**Reply by srstsavage** [reply]
Converted to #552

**Reply by NiKiZe** [reply]
I do not think that an Issue is the correct here since the only issue with iPXE would probably be missing native driver, but that will not really be covered by an issue.

There is no driver in iPXE for any `8020` device.
Have you tried to contact QLogic about this?

**Reply by srstsavage** [reply]
> Have you tried to contacted QLogic about this?

I haven't, but will do so

**Reply by NiKiZe** [reply]
Is this still relevant hardware?
We have ifstat from undionly.kpxe, which can only be undi, but what is the driver from ipxe.pxe?
Any chance to try ipxe.efi?

In undi, yes the nic rom is the issue, and not much iPXE will be able to do.
So the issue #552 was closed, mostly as I don't think this is still relevant (but happy to be wrong) but also since I don't see that iPXE can do anything in the undionly case, but potentially in the ipxe.pxe case.
